rfcs/rfc-covid-agm-extension.md

```diff
@@ -0,0 +1,48 @@
+# RFC For COVID AGM date extension
+
+- Start Date: (2021-01-19)
+- Target Major Version: (None)
+- Reference Issues: (None)
+- Entity Issue: (https://app.zenhub.com/workspaces/entity-5bf2f2164b5806bc2bf60531/issues/bcgov/entity/6198)
+- Implementation PR: (None)
+
+# Summary
+
+The registrar gave all co-ops an extension on their AGMs for COVID. We need to enable them to select those extended dates in the AR with the AGM date picker. It currently restricts them to Jan 1st - Dec 31st of the calendar year.
+
+First, we need to update the default date picker validation, and allow an AGM date anywhere from Jan 1st to April 30th of the next year. Currently it is Jan 1st to Dec 31st of the same year. This is because many co-ops filing their 2020 AR hold the AGM for that year in 2021
+
+Our solution for the COVID AGM extensions is to have them check a checkbox that they are taking the blanket extension. When they check the checkbox, validation on the date picker will go from the default up to April 30th of the next year, to 6 months ahead (October 31st of the next year).
+
+ - This COVID AGM extension checkbox is optional.
+ - It's only purpose is to extend the AGM date picker validation.
+ - It will only exist in the 2020 Annual Report filing.
+ - There is no change to the ledger item if the checkbox is selected.
+ - We do not need to store if the check box was checked or not for later reference.
+
+# Basic example
+
+# Motivation
+
+# Detailed design
+
+Changes required in annual_report validation and get todo task.
+  - In validation `expected_date` has been calculated from `last_ar_date` by adding a year, which creates an issue with the new requirement
+
+    Example: Filing an ar for 2020 on 2021 Aug 15th the `last_filing_ar` will set to 2021 Aug 15th, when user tries to file an ar for the year 2021 the existing validation restrict to file on or before 2022 Aug 15th. Based on the requirement user cannot file after 2022 Apr 30. For the first time filer has same kind of issue. If the founding date is 2020 Jan 15th and trying to file on Mar 15th, the validation restrict to file after 2021 Jan 15th
+
+  - In todo task `todo_start_date` is calculating in the same logic as above (by adding 1 year). This will fail in the upcomming year.
+
```

Comment:
Please also mention the "min date" and "max date" rules for a Coop AR filing. Ie,

```
  private get minDate (): string {
    const firstDayOfYear = `${this.ARFilingYear}-01-01`
    const lastAnnualReportDate = this.lastAnnualReportDate
    // min date is the AR year on Jan 1
    // or the date of the previous AR (in case of 2 ARs filed in the same year)
    // whichever is latest
    return this.latestDate(firstDayOfYear, lastAnnualReportDate)
  }

  private get maxDate (): string {
    if (this.ARFilingYear === 2020) {
      // special case for 2020 ARs!
      // max date is today or Oct 31, 2021, whichever is earliest
      // NB: UI still needs to decide whether to display extension message
      // NB: UI still needs to decide whether date is valid
      return this.earliestDate(this.currentDate, '2021-10-31')
    } else if (this.ARFilingYear < this.currentYear) {
      // for past ARs, max date is the following year on Apr 30
      return `${this.ARFilingYear + 1}-04-30`
    } else {
      // for current ARs, max date is today
      return this.currentDate
    }
  }
```

Reply:
Please also mention that the API will provide the Coop AR min date and max date in the todo task. Ie, in here:

![image](https://user-images.githubusercontent.com/10002889/105219562-732c8f00-5b0b-11eb-8efd-60b1eb98c716.png)

Reply:
Please also mention that there is no code change needed (in businesses endpoint or the tasks endpoint) for BCOMP ARs.

Reply:
Please also mention our assumption that the "as of date", if the user selects "we did not hold an AGM", is Dec 31 of the AR year (same as now).